src/mam/mod_mam_utils.erl

```diff
@@ -390,6 +391,33 @@ has_chat_marker(Packet) ->
         _                                 -> false
     end.
 
+get_retract_id(Packet) ->
+    case exml_query:subelement_with_ns(Packet, ?NS_FASTEN) of
+        El = #xmlel{name = <<"apply-to">>} ->
+            case exml_query:subelement_with_ns(El, ?NS_RETRACT) of
+                #xmlel{name = <<"retract">>} -> exml_query:attr(El, <<"id">>, none);
+                _ -> none
+            end;
+        _ -> none
+    end.
```

Comment:
There is `exml:subelement_with_name_and_ns/3` which returns an element of the given name only if the namespaces attributes match. Also, the whole id extraction could be done with exml query like this one:
```erlang
exml_query:path(Packet, [{element_with_ns, <<"apply-to">>, ?NS_FASTEN},
                         {element_with_ns, <<"retract">>, ?NS_RETRACT},
                         {attr, <<"id">>}], none)
```

That's just a hint. The current code looks ok.

Reply:
Yes, you're right, I forgot to rework this function, I only changed `get_origin_id` like this after some code changes.

Reply:
Actually no, you're wrong ;-)
The attribute is taken from the `apply-to` element, not from the `retract` one, so your code would not work. I could use `exml_query:subelement_with_name_and_ns/3` here though.